File: renderer/blender/fetch_google_tiles.py
#!/usr/bin/env python3
"""
Blender Python script to fetch Google 3D Tiles using the Blosm addon.
Downloads photorealistic 3D tiles for a specified area and exports to OBJ + textures.

Prerequisites:
    - Blender 4.3.2+ with Blosm addon installed
    - Google Maps Tiles API key (unrestricted)
    - Maps Tiles API enabled in Google Cloud Console

Usage:
    blender --background --python fetch_google_tiles.py -- config.json
"""

import bpy
import json
import sys
import os
import math

# Default configuration
DEFAULT_CONFIG = {
    "minLon": -83.026,
    "maxLon": -83.012,
    "minLat": 39.996,
    "maxLat": 40.006,
    "apiKey": None,
    "outputDir": None,
    "detailLevel": "medium",  # "low", "medium", "high"
    "exportFormat": "obj",  # "obj", "gltf", "fbx"
}

# Detail level mapping for Blosm
DETAIL_LEVELS = {
    "low": "city",
    "medium": "buildings",
    "high": "buildings_more_details",
}


def check_blosm_addon():
    """Check if Blosm addon is installed and enabled."""
    addon_name = "blosm"

    # Check if addon is in preferences
    if addon_name not in bpy.context.preferences.addons:
        # Try to enable it
        try:
            bpy.ops.preferences.addon_enable(module=addon_name)
            print(f"Enabled {addon_name} addon")
        except Exception as e:
            print(f"ERROR: Blosm addon not found or could not be enabled: {e}")
            print("\nTo install Blosm:")
            print("  1. Download from: https://prochitecture.gumroad.com/l/blosm")
            print("  2. In Blender: Edit > Preferences > Add-ons > Install")
            print("  3. Select the downloaded .zip file")
            print("  4. Enable the 'blosm' addon")
            return False

    return True


def setup_blosm_settings(api_key, bounds, detail_level):
    """Configure Blosm addon settings for Google 3D Tiles."""
    try:
        # Access Blosm preferences
        addon_prefs = bpy.context.preferences.addons.get("blosm")
        if addon_prefs and hasattr(addon_prefs, "preferences"):
            prefs = addon_prefs.preferences

            # Set API key
            if hasattr(prefs, "googleMapsApiKey"):
                prefs.googleMapsApiKey = api_key
            elif hasattr(prefs, "google3dTilesApiKey"):
                prefs.google3dTilesApiKey = api_key

            print(f"Configured Blosm with API key: {api_key[:10]}...")

        # Access scene-level Blosm properties
        scene = bpy.context.scene
        if hasattr(scene, "blosm"):
            blosm = scene.blosm

            # Set data source to Google 3D Tiles
            if hasattr(blosm, "dataType"):
                blosm.dataType = "3d_tiles"

            if hasattr(blosm, "source"):
                blosm.source = "google"

            # Set coordinates
            if hasattr(blosm, "minLon"):
                blosm.minLon = bounds["minLon"]
                blosm.maxLon = bounds["maxLon"]
                blosm.minLat = bounds["minLat"]
                blosm.maxLat = bounds["maxLat"]

            # Set detail level
            if hasattr(blosm, "lodLevel"):
                blosm.lodLevel = DETAIL_LEVELS.get(detail_level, "buildings")

            print(f"Set bounds: [{bounds['minLon']}, {bounds['minLat']}] to [{bounds['maxLon']}, {bounds['maxLat']}]")
            print(f"Detail level: {detail_level}")

        return True

    except Exception as e:
        print(f"ERROR: Failed to configure Blosm: {e}")
        return False


def import_google_tiles():
    """Trigger Blosm import of Google 3D Tiles."""
    try:
        # The Blosm operator for importing 3D tiles
        if hasattr(bpy.ops, "blosm") and hasattr(bpy.ops.blosm, "import_data"):
            bpy.ops.blosm.import_data()
            print("Import triggered successfully")
            return True
        elif hasattr(bpy.ops, "import_scene") and hasattr(bpy.ops.import_scene, "blosm"):
            bpy.ops.import_scene.blosm()
            print("Import triggered successfully")
            return True
        else:
            print("ERROR: Could not find Blosm import operator")
            print("Available blosm operators:", dir(bpy.ops.blosm) if hasattr(bpy.ops, "blosm") else "None")
            return False

    except Exception as e:
        print(f"ERROR: Import failed: {e}")
        return False


def export_scene(output_dir, export_format, base_name="google_tiles"):
    """Export the imported 3D tiles to specified format."""
    os.makedirs(output_dir, exist_ok=True)

    # Select all mesh objects
    bpy.ops.object.select_all(action='DESELECT')
    for obj in bpy.data.objects:
        if obj.type == 'MESH':
            obj.select_set(True)

    mesh_count = len([o for o in bpy.data.objects if o.type == 'MESH'])
    print(f"Exporting {mesh_count} mesh objects...")

    if mesh_count == 0:
        print("WARNING: No mesh objects to export")
        return None

    if export_format == "obj":
        filepath = os.path.join(output_dir, f"{base_name}.obj")
        bpy.ops.wm.obj_export(
            filepath=filepath,
            export_selected_objects=True,
            export_materials=True,
            export_uv=True,
            export_normals=True,
        )
        print(f"Exported OBJ: {filepath}")

    elif export_format == "gltf":
        filepath = os.path.join(output_dir, f"{base_name}.glb")
        bpy.ops.export_scene.gltf(
            filepath=filepath,
            use_selection=True,
            export_format='GLB',
            export_materials='EXPORT',
            export_texcoords=True,
        )
        print(f"Exported glTF: {filepath}")

    elif export_format == "fbx":
        filepath = os.path.join(output_dir, f"{base_name}.fbx")
        bpy.ops.export_scene.fbx(
            filepath=filepath,
            use_selection=True,
            embed_textures=True,
        )
        print(f"Exported FBX: {filepath}")

    # Also export textures separately
    export_textures(output_dir)

    return filepath


def export_textures(output_dir):
    """Unpack and save all textures to output directory."""
    textures_dir = os.path.join(output_dir, "textures")
    os.makedirs(textures_dir, exist_ok=True)

    # Unpack all images
    texture_count = 0
    for image in bpy.data.images:
        if image.packed_file:
            try:
                # Set filepath and unpack
                image.filepath_raw = os.path.join(textures_dir, image.name)
                image.save()
                texture_count += 1
            except Exception as e:
                print(f"Warning: Could not save texture {image.name}: {e}")

    # Also try the built-in unpack
    try:
        bpy.ops.file.unpack_all(method='WRITE_LOCAL')
    except:
        pass

    print(f"Exported {texture_count} textures to: {textures_dir}")


def write_metadata(output_dir, config, stats):
    """Write metadata about the imported tiles."""
    metadata = {
        "source": "Google Photorealistic 3D Tiles",
        "bounds": {
            "minLon": config["minLon"],
            "maxLon": config["maxLon"],
            "minLat": config["minLat"],
            "maxLat": config["maxLat"],
        },
        "detailLevel": config["detailLevel"],
        "exportFormat": config["exportFormat"],
        "stats": stats,
        "attribution": "Map data ©2024 Google",
    }

    metadata_path = os.path.join(output_dir, "google-tiles-metadata.json")
    with open(metadata_path, 'w') as f:
        json.dump(metadata, f, indent=2)

    print(f"Metadata written: {metadata_path}")


def get_scene_stats():
    """Get statistics about the imported scene."""
    mesh_count = 0
    vertex_count = 0
    face_count = 0

    for obj in bpy.data.objects:
        if obj.type == 'MESH':
            mesh_count += 1
            vertex_count += len(obj.data.vertices)
            face_count += len(obj.data.polygons)

    return {
        "meshCount": mesh_count,
        "vertexCount": vertex_count,
        "faceCount": face_count,
        "textureCount": len(bpy.data.images),
        "materialCount": len(bpy.data.materials),
    }


def clear_scene():
    """Remove all objects from scene."""
    bpy.ops.object.select_all(action='SELECT')
    bpy.ops.object.delete(use_global=False)


def main():
    """Main entry point."""
    print("=" * 60)
    print("Google 3D Tiles Fetcher (Blosm)")
    print("=" * 60)

    # Parse arguments
    argv = sys.argv
    if "--" in argv:
        args = argv[argv.index("--") + 1:]
    else:
        args = []

    # Load config
    if len(args) < 1:
        print("Usage: blender --background --python fetch_google_tiles.py -- config.json")
        sys.exit(1)

    config_path = args[0]
    print(f"Loading config: {config_path}")

    with open(config_path, 'r') as f:
        config = {**DEFAULT_CONFIG, **json.load(f)}

    # Validate required fields
    if not config.get("apiKey"):
        print("ERROR: apiKey is required in config")
        print("Get a free API key at: https://console.cloud.google.com/")
        sys.exit(1)

    if not config.get("outputDir"):
        print("ERROR: outputDir is required in config")
        sys.exit(1)

    # Check Blosm addon
    print("\nChecking Blosm addon...")
    if not check_blosm_addon():
        sys.exit(1)

    # Clear scene
    print("\nClearing scene...")
    clear_scene()

    # Configure Blosm
    print("\nConfiguring Blosm...")
    bounds = {
        "minLon": config["minLon"],
        "maxLon": config["maxLon"],
        "minLat": config["minLat"],
        "maxLat": config["maxLat"],
    }

    if not setup_blosm_settings(config["apiKey"], bounds, config["detailLevel"]):
        sys.exit(1)

    # Import tiles
    print("\nImporting Google 3D Tiles...")
    print("This may take several minutes depending on area size and detail level...")

    if not import_google_tiles():
        print("\n" + "=" * 60)
        print("MANUAL IMPORT REQUIRED")
        print("=" * 60)
        print("\nBlosm automated import is not available in background mode.")
        print("Please follow these steps manually:")
        print("\n1. Open Blender (GUI mode)")
        print("2. Press N to open the sidebar, find the Blosm panel")
        print("3. Select '3D Tiles' from the dropdown")
        print("4. Set Source to 'Google'")
        print(f"5. Enter coordinates:")
        print(f"   Min: {bounds['minLon']}, {bounds['minLat']}")
        print(f"   Max: {bounds['maxLon']}, {bounds['maxLat']}")
        print("6. Click 'Import'")
        print(f"7. Export to: {config['outputDir']}")
        print("=" * 60)

        # Save a Blender file with settings pre-configured for manual import
        blend_path = os.path.join(config["outputDir"], "google_tiles_import.blend")
        os.makedirs(config["outputDir"], exist_ok=True)
        bpy.ops.wm.save_as_mainfile(filepath=blend_path)
        print(f"\nSaved Blender file for manual import: {blend_path}")

        sys.exit(0)

    # Get stats
    stats = get_scene_stats()
    print(f"\nImport complete!")
    print(f"  Meshes: {stats['meshCount']}")
    print(f"  Vertices: {stats['vertexCount']:,}")
    print(f"  Faces: {stats['faceCount']:,}")
    print(f"  Textures: {stats['textureCount']}")

    # Export
    print(f"\nExporting to {config['exportFormat'].upper()}...")
    export_scene(config["outputDir"], config["exportFormat"])

    # Write metadata
    write_metadata(config["outputDir"], config, stats)

    print("\nDONE!")


if __name__ == "__main__":
    main()

File: renderer/scripts/fetch-google-tiles.ts
#!/usr/bin/env node
/**
 * Fetch Google Photorealistic 3D Tiles for the project bounds.
 * Uses Blosm addon in Blender to download and export tiles.
 *
 * Prerequisites:
 *   - Blender 4.3.2+ with Blosm addon installed
 *   - GOOGLE_MAPS_API_KEY environment variable (or in .env)
 *   - Maps Tiles API enabled in Google Cloud Console
 *
 * Get API key: https://console.cloud.google.com/google/maps-apis/credentials
 * Install Blosm: https://prochitecture.gumroad.com/l/blosm
 */

import { spawn, execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Paths
const RENDERER_ROOT = path.resolve(__dirname, '..', '..');
const PROJECT_ROOT = path.resolve(RENDERER_ROOT, '..');
const BLENDER_SCRIPT = path.join(RENDERER_ROOT, 'blender', 'fetch_google_tiles.py');
const OUTPUT_DIR = path.join(PROJECT_ROOT, 'data', 'google-tiles');

// Focus bounds (same as other scripts)
const FOCUS_BOUNDS = {
  minLon: -83.026,
  maxLon: -83.012,
  minLat: 39.996,
  maxLat: 40.006,
};

interface FetchConfig {
  minLon: number;
  maxLon: number;
  minLat: number;
  maxLat: number;
  apiKey: string;
  outputDir: string;
  detailLevel: 'low' | 'medium' | 'high';
  exportFormat: 'obj' | 'gltf' | 'fbx';
}

/**
 * Load Google Maps API key from environment or .env file.
 */
function getApiKey(): string | null {
  // Check environment variable
  if (process.env.GOOGLE_MAPS_API_KEY) {
    return process.env.GOOGLE_MAPS_API_KEY;
  }

  // Try loading from .env file
  const envPath = path.join(PROJECT_ROOT, '.env');
  if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf-8');
    const match = envContent.match(/GOOGLE_MAPS_API_KEY=(.+)/);
    if (match) {
      return match[1].trim();
    }
  }

  return null;
}

/**
 * Find Blender executable path.
 */
function findBlender(): string | null {
  const platform = os.platform();
  const paths: string[] = [];

  if (platform === 'darwin') {
    paths.push('/Applications/Blender.app/Contents/MacOS/Blender');
    paths.push('/Applications/Blender.app/Contents/MacOS/blender');
  } else if (platform === 'win32') {
    paths.push('C:\\Program Files\\Blender Foundation\\Blender 4.3\\blender.exe');
    paths.push('C:\\Program Files\\Blender Foundation\\Blender 4.2\\blender.exe');
    paths.push('C:\\Program Files\\Blender Foundation\\Blender 4.0\\blender.exe');
  } else {
    paths.push('/usr/bin/blender');
    paths.push('/usr/local/bin/blender');
    paths.push('/snap/bin/blender');
  }

  for (const p of paths) {
    if (fs.existsSync(p)) {
      return p;
    }
  }

  // Try 'which' or 'where' command
  try {
    const cmd = platform === 'win32' ? 'where blender' : 'which blender';
    const result = execSync(cmd, { encoding: 'utf-8' }).trim();
    if (result && fs.existsSync(result.split('\n')[0])) {
      return result.split('\n')[0];
    }
  } catch {
    // Command failed
  }

  return null;
}

/**
 * Check Blender version.
 */
function checkBlenderVersion(blenderPath: string): string | null {
  try {
    const result = execSync(`"${blenderPath}" --version`, { encoding: 'utf-8' });
    const match = result.match(/Blender (\d+\.\d+)/);
    return match ? match[1] : null;
  } catch {
    return null;
  }
}

/**
 * Run Blender with the fetch script.
 */
async function runBlender(blenderPath: string, configPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const args = [
      '--background',
      '--python', BLENDER_SCRIPT,
      '--',
      configPath,
    ];

    console.log(`Running: ${blenderPath} ${args.join(' ')}`);

    const proc = spawn(blenderPath, args, {
      stdio: ['ignore', 'pipe', 'pipe'],
    });

    proc.stdout.on('data', (data: Buffer) => {
      const lines = data.toString().split('\n');
      for (const line of lines) {
        if (line.trim()) {
          console.log(`  ${line}`);
        }
      }
    });

    proc.stderr.on('data', (data: Buffer) => {
      const text = data.toString().trim();
      if (text && !text.includes('Fra:') && !text.includes('Mem:')) {
        console.error(`  stderr: ${text}`);
      }
    });

    proc.on('close', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(`Blender exited with code ${code}`));
      }
    });

    proc.on('error', (err) => {
      reject(new Error(`Failed to spawn Blender: ${err.message}`));
    });
  });
}

/**
 * Print manual instructions for Blosm import.
 */
function printManualInstructions(config: FetchConfig): void {
  console.log('\n' + '='.repeat(60));
  console.log('MANUAL IMPORT INSTRUCTIONS');
  console.log('='.repeat(60));
  console.log('\nBlosm requires GUI mode for initial setup. Follow these steps:\n');

  console.log('1. INSTALL BLOSM ADDON:');
  console.log('   - Download from: https://prochitecture.gumroad.com/l/blosm');
  console.log('   - In Blender: Edit > Preferences > Add-ons > Install');
  console.log('   - Enable the "blosm" addon\n');

  console.log('2. CONFIGURE API KEY:');
  console.log('   - In Blender: Edit > Preferences > Add-ons > blosm');
  console.log(`   - Enter your Google Maps API key: ${config.apiKey.substring(0, 15)}...`);
  console.log('   - Click "Save Preferences"\n');

  console.log('3. IMPORT 3D TILES:');
  console.log('   - Press N to open sidebar, find "Blosm" panel');
  console.log('   - Select "3D Tiles" from dropdown');
  console.log('   - Set Source to "Google"');
  console.log('   - Click "select" to open map, or enter coordinates:');
  console.log(`     Min Longitude: ${config.minLon}`);
  console.log(`     Max Longitude: ${config.maxLon}`);
  console.log(`     Min Latitude: ${config.minLat}`);
  console.log(`     Max Latitude: ${config.maxLat}`);
  console.log(`   - Detail Level: ${config.detailLevel}`);
  console.log('   - Click "Import"\n');

  console.log('4. EXPORT:');
  console.log('   - File > Export > glTF 2.0 (.glb/.gltf)');
  console.log(`   - Save to: ${config.outputDir}/google_tiles.glb\n`);

  console.log('5. EXTRACT TEXTURES:');
  console.log('   - In Image Editor, save each texture');
  console.log('   - Or: File > External Data > Unpack Resources\n');

  console.log('='.repeat(60));
}

async function main() {
  console.log('Google 3D Tiles Fetcher');
  console.log('='.repeat(50));

  // Parse arguments
  const args = process.argv.slice(2);

  // Get detail level
  let detailLevel: 'low' | 'medium' | 'high' = 'medium';
  const detailIdx = args.indexOf('--detail');
  if (detailIdx !== -1 && args[detailIdx + 1]) {
    const level = args[detailIdx + 1].toLowerCase();
    if (['low', 'medium', 'high'].includes(level)) {
      detailLevel = level as 'low' | 'medium' | 'high';
    }
  }

  // Get export format
  let exportFormat: 'obj' | 'gltf' | 'fbx' = 'gltf';
  const formatIdx = args.indexOf('--format');
  if (formatIdx !== -1 && args[formatIdx + 1]) {
    const format = args[formatIdx + 1].toLowerCase();
    if (['obj', 'gltf', 'fbx'].includes(format)) {
      exportFormat = format as 'obj' | 'gltf' | 'fbx';
    }
  }

  // Check for manual mode
  const manualMode = args.includes('--manual');

  console.log(`Detail level: ${detailLevel}`);
  console.log(`Export format: ${exportFormat}`);

  // Get API key
  const apiKey = getApiKey();
  if (!apiKey) {
    console.error('\nERROR: GOOGLE_MAPS_API_KEY not found!');
    console.error('\nTo get an API key:');
    console.error('  1. Go to: https://console.cloud.google.com/google/maps-apis/credentials');
    console.error('  2. Create a new API key (unrestricted)');
    console.error('  3. Enable "Map Tiles API" in your project');
    console.error('  4. Add to .env file: GOOGLE_MAPS_API_KEY=your_key_here');
    process.exit(1);
  }
  console.log(`API Key: ${apiKey.substring(0, 15)}...`);

  // Build config
  const config: FetchConfig = {
    ...FOCUS_BOUNDS,
    apiKey,
    outputDir: OUTPUT_DIR,
    detailLevel,
    exportFormat,
  };

  // Ensure output directory exists
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // If manual mode, just print instructions
  if (manualMode) {
    printManualInstructions(config);
    process.exit(0);
  }

  // Find Blender
  console.log('\nLocating Blender...');
  const blenderPath = findBlender();
  if (!blenderPath) {
    console.error('ERROR: Blender not found!');
    console.error('Please install Blender 4.3+ from https://www.blender.org/download/');
    process.exit(1);
  }
  console.log(`Found: ${blenderPath}`);

  // Check version
  const version = checkBlenderVersion(blenderPath);
  if (version) {
    console.log(`Version: ${version}`);
    const major = parseFloat(version);
    if (major < 4.3) {
      console.warn(`Warning: Blender ${version} detected. Version 4.3+ recommended for texture unpacking.`);
    }
  }

  // Write config to temp file
  const configPath = path.join(os.tmpdir(), `google-tiles-config-${Date.now()}.json`);
  fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
  console.log(`Config written to: ${configPath}`);

  // Calculate area info
  const latCenter = (config.minLat + config.maxLat) / 2;
  const lonSpan = (config.maxLon - config.minLon) * 111320 * Math.cos(latCenter * Math.PI / 180);
  const latSpan = (config.maxLat - config.minLat) * 111320;
  console.log(`\nArea: ${(lonSpan / 1000).toFixed(2)} km × ${(latSpan / 1000).toFixed(2)} km`);
  console.log(`Center: ${latCenter.toFixed(4)}, ${((config.minLon + config.maxLon) / 2).toFixed(4)}`);

  // Run Blender
  console.log('\nStarting Blender...');
  console.log('Note: Blosm may require GUI mode for first-time setup.\n');

  try {
    await runBlender(blenderPath, configPath);
    console.log('\nFetch complete!');
    console.log(`Output: ${OUTPUT_DIR}`);

    // Check what was exported
    if (fs.existsSync(path.join(OUTPUT_DIR, 'google_tiles.glb'))) {
      console.log('  - google_tiles.glb');
    }
    if (fs.existsSync(path.join(OUTPUT_DIR, 'google_tiles.obj'))) {
      console.log('  - google_tiles.obj');
    }
    if (fs.existsSync(path.join(OUTPUT_DIR, 'textures'))) {
      const textures = fs.readdirSync(path.join(OUTPUT_DIR, 'textures'));
      console.log(`  - textures/ (${textures.length} files)`);
    }

  } catch (err) {
    console.error('\nBlender script failed.');
    console.log('\nThis is expected if Blosm addon is not installed or configured.');
    printManualInstructions(config);
    process.exit(1);
  } finally {
    // Clean up temp config
    try {
      fs.unlinkSync(configPath);
    } catch {
      // Ignore
    }
  }
}

main().catch((err) => {
  console.error('Error:', err);
  process.exit(1);
});
